src/main/java/io/netty/incubator/codec/http3/Http3ServerPushStreamManager.java

```diff
@@ -0,0 +1,280 @@
+import java.util.function.UnaryOperator;
+
+import static io.netty.incubator.codec.http3.Http3.maxPushIdReceived;
+import static io.netty.incubator.codec.http3.Http3CodecUtils.connectionError;
+import static io.netty.incubator.codec.http3.Http3ErrorCode.H3_ID_ERROR;
+import static io.netty.util.internal.PlatformDependent.newConcurrentHashMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+/**
+ * A manager for <a href="https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-push-streams">push streams</a>
+ * for a server. New push streams can be initiated using the various {@code newPushStream} methods. It is required to
+ * add the {@link ChannelHandler} returned from {@link #controlStreamListener()} to the {@link QuicChannel} associated
+ * with this manager.
+ */
+public final class Http3ServerPushStreamManager {
+    private static final AtomicLongFieldUpdater<Http3ServerPushStreamManager> nextIdUpdater =
+            newUpdater(Http3ServerPushStreamManager.class, "nextId");
+    private static final Object CANCELLED_STREAM = new Object();
+    private static final Object PUSH_ID_GENERATED = new Object();
+    private static final Object AWAITING_STREAM_ESTABLISHMENT = new Object();
+
+    private final QuicChannel channel;
+    private final ConcurrentMap<Long, Object> pushStreams;
+    private final ChannelInboundHandler controlStreamListener;
+
+    private volatile long nextId;
+
+    public Http3ServerPushStreamManager(QuicChannel channel) {
+        this(channel, 8);
+    }
+
+    public Http3ServerPushStreamManager(QuicChannel channel, int initialPushStreamsCountHint) {
+        this.channel = requireNonNull(channel);
+        pushStreams = newConcurrentHashMap(initialPushStreamsCountHint);
+        controlStreamListener = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                if (msg instanceof Http3CancelPushFrame) {
+                    final long pushId = ((Http3CancelPushFrame) msg).id();
+                    if (pushId >= nextId) {
+                        connectionError(ctx, H3_ID_ERROR, "CANCEL_PUSH id greater than the last known id", true);
+                        return;
+                    }
+
+                    pushStreams.computeIfPresent(pushId, (id, existing) -> {
+                        if (existing == AWAITING_STREAM_ESTABLISHMENT) {
+                            return CANCELLED_STREAM;
+                        }
+                        if (existing == PUSH_ID_GENERATED) {
+                            throw new IllegalStateException("Unexpected push stream state " + existing +
+                                    " for pushId: " + id);
+                        }
+                        assert existing instanceof QuicStreamChannel;
+                        ((QuicStreamChannel) existing).close();
```

Comment:
nit: this could just be `Channel`

Reply:
Since we have to cast anyway and we know that it should be a `QuicStreamChannel`, I think it is better to just cast to `QuicStreamChannel`.